In [ ]:
import pandas as pd
import numpy as np
from sklearn.impute import KNNImputer
from sklearn.preprocessing import LabelEncoder

In [ ]:
features = pd.read_csv("/content/features.csv")
labels = pd.read_csv("/content/labels.csv")

**1️. Tampilkan data yang telah diunduh, kemudian jelaskan semua kolom yang terdapat pada dataset tersebut.**




In [ ]:
features.head().T


0           1  \
Unnamed: 0                                               0           1   
Student Id                                            6634        6459   
Marital status                                      single      single   
Application mode                                       1.0         1.0   
Application order                                      3.0         1.0   
Course                                              9773.0         NaN   
Daytime/evening attendance\t                           day         day   
Previous qualification                                 1.0         1.0   
Previous qualification (grade)                         NaN       136.0   
Nacionality                                     Portuguese  Portuguese   
Mother's qualification                                 1.0        19.0   
Father's qualification                                 1.0        19.0   
Mother's occupation                                    4.0         4.0   
Father's occupation                                    7.0         8.0   
Admission grade                                      125.1       150.4   
Displaced                                              1.0         1.0   
Educational special needs                              0.0         0.0   
Debtor                                                 1.0         0.0   
Tuition fees up to date                                1.0         1.0   
Gender                                                 1.0         1.0   
Scholarship holder                                     NaN         1.0   
Age at enrollment                                     18.0        21.0   
International                                          0.0         0.0   
Curricular units 1st sem (credited)                    0.0         0.0   
Curricular units 1st sem (enrolled)                    6.0         0.0   
Curricular units 1st sem (evaluations)                 6.0         0.0   
Curricular units 1st sem (approved)                    6.0         0.0   
Curricular units 1st sem (grade)                 14.333333         0.0   
Curricular units 1st sem (without evaluations)         0.0         0.0   
Curricular units 2nd sem (credited)                    0.0         0.0   
Curricular units 2nd sem (enrolled)                    6.0         0.0   
Curricular units 2nd sem (evaluations)                 6.0         0.0   
Curricular units 2nd sem (approved)                    6.0         0.0   
Curricular units 2nd sem (grade)                 15.333333         0.0   
Curricular units 2nd sem (without evaluations)         0.0         0.0   
Unemployment rate                                     11.1        12.4   
Inflation rate                                         0.6         0.5   
GDP                                                   2.02        1.79   

                                                         2           3  \
Unnamed: 0                                               2           3   
Student Id                                            2238        2479   
Marital status                                      single      single   
Application mode                                      42.0         1.0   
Application order                                      1.0         1.0   
Course                                              9500.0         NaN   
Daytime/evening attendance\t                           day         day   
Previous qualification                                 1.0         1.0   
Previous qualification (grade)                       120.0       141.0   
Nacionality                                     Portuguese  Portuguese   
Mother's qualification                                 1.0         1.0   
Father's qualification                                 2.0        19.0   
Mother's occupation                                    9.0         4.0   
Father's occupation                                    3.0         4.0   
Admission grade                                      130.0  

In [ ]:
labels.head()

Target
0  Graduate
1  Graduate
2  Graduate
3  Graduate
4   Dropout

Penjelasan Kolom Dataset
Dataset ini terdiri dari beberapa kolom yang memberikan gambaran tentang mahasiswa. Berikut penjelasan setiap kolom:

Unnamed: 0 : Nomor urut atau indeks otomatis dari dataset.

Student Id : ID unik untuk setiap mahasiswa.

Marital status : Status pernikahan mahasiswa (misalnya single, married, dll).

Application mode : Jalur pendaftaran yang digunakan mahasiswa (tes nasional, transfer, jalur mandiri, dll).

Application order : Urutan pilihan program studi saat mahasiswa mendaftar.

Course : Program studi yang dipilih oleh mahasiswa.

Daytime/evening attendance : Waktu kuliah mahasiswa (day = kelas pagi, evening = kelas malam).

Previous qualification : Kualifikasi pendidikan terakhir sebelum masuk universitas.

Previous qualification (grade) : Nilai atau rata-rata nilai pada pendidikan sebelumnya.

Nationality : Kewarganegaraan mahasiswa.

Mother’s qualification : Tingkat pendidikan terakhir ibu mahasiswa.

Father’s qualification : Tingkat pendidikan terakhir ayah mahasiswa.

Mother’s occupation : Pekerjaan ibu mahasiswa.

Father’s occupation : Pekerjaan ayah mahasiswa.

Admission grade : Nilai yang diperoleh saat seleksi masuk universitas.

Displaced : Menunjukkan apakah mahasiswa berasal dari daerah pengungsian/terpencil (1 = ya, 0 = tidak).

Educational special needs : Menunjukkan apakah mahasiswa memiliki kebutuhan khusus (1 = ya, 0 = tidak).

Debtor : Menunjukkan apakah mahasiswa memiliki tunggakan pembayaran (1 = ya, 0 = tidak).

Tuition fees up to date : Apakah mahasiswa sudah melunasi pembayaran biaya kuliah (1 = ya, 0 = tidak).

Gender : Jenis kelamin mahasiswa (male/female).

Scholarship holder : Menunjukkan apakah mahasiswa menerima beasiswa (1 = ya, 0 = tidak).

Age at enrollment : Usia mahasiswa saat mendaftar.

International : Menunjukkan apakah mahasiswa tersebut merupakan mahasiswa internasional (1 = ya, 0 = tidak).

Curricular units 1st sem (credited) : Jumlah mata kuliah semester 1 yang diakui.

Curricular units 1st sem (enrolled) : Jumlah mata kuliah semester 1 yang diambil.

Curricular units 1st sem (evaluations) : Jumlah evaluasi pada semester 1.

Curricular units 1st sem (approved) : Jumlah mata kuliah semester 1 yang lulus.

Curricular units 1st sem (grade) : Rata-rata nilai semester 1.

Curricular units 1st sem (without evaluations) : Jumlah mata kuliah semester 1 tanpa evaluasi.

Curricular units 2nd sem (credited) : Jumlah mata kuliah semester 2 yang diakui.

Curricular units 2nd sem (enrolled) : Jumlah mata kuliah semester 2 yang diambil.

Curricular units 2nd sem (evaluations) : Jumlah evaluasi pada semester 2.

Curricular units 2nd sem (approved) : Jumlah mata kuliah semester 2 yang lulus.

Curricular units 2nd sem (grade) : Rata-rata nilai semester 2.

Curricular units 2nd sem (without evaluations) : Jumlah mata kuliah semester 2 tanpa evaluasi.

Unemployment rate : Tingkat pengangguran di tahun pendaftaran.

Inflation rate : Tingkat inflasi pada tahun pendaftaran.

GDP : Nilai Produk Domestik Bruto nasional pada tahun pendaftaran.

Penjelasan Kolom Dataset Labels

Kolom pada dataset labels berisi informasi status akhir mahasiswa:

Graduate : Mahasiswa telah lulus.

Dropout : Mahasiswa berhenti kuliah sebelum lulus.

Enrolled : Mahasiswa masih aktif kuliah.


**2️. Lakukan agregasi antara tabel features dengan tabel labels.**



In [ ]:
df = pd.concat([features, labels], axis=1)
df

Unnamed: 0  Student Id Marital status  Application mode  \
0              0        6634         single               1.0   
1              1        6459         single               1.0   
2              2        2238         single              42.0   
3              3        2479         single               1.0   
4              4        4260         single              44.0   
...          ...         ...            ...               ...   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
0                   3.0  9773.0                          day   
1                   1.0     NaN                          day   
2                   1.0  9500.0                          day   
3                   1.0     NaN                          day   
4                   1.0     NaN                          day   
...                 ...     ...                          ...   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                        1.0                             NaN  Portuguese  ...   
1                        1.0                           136.0  Portuguese  ...   
2                        1.0                           120.0  Portuguese  ...   
3                        1.0                           141.0  Portuguese  ...   
4                       39.0                           120.0     Spanish  ...   
...                      ...                             ...         ...  ...   
3091                     1.0                           140.0  Portuguese  ...   
3092                     1.0                           126.0  Portuguese  ...   
3093                     1.0                           141.0  Portuguese  ...   
3094                     4.0                           150.0  Portuguese  ...   
3095                     1.0                           118.0  Portuguese  ...   

      Curricular units 2nd sem (credited)  \
0                                     0.0   
1                                     0.0   
2                                     4.0   
3                                     0.0   
4                                     1.0   
...                                   ...   
3091                                  1.0   
3092                                  0.0   
3093                                  0.0   
3094                                  2.0   
3095                                  0.0   

      Curricular units 2nd sem (enrolled)  \
0                                     6.0   
1                                     0.0   
2                                     8.0   
3                                     6.0   
4                                     6.0   
...                                   ...   
3091                                  6.0   
3092                                  6.0   
3093                                  8.0   
3094                                  8.0   
3095                                  6.0   

      Curricular units 2nd sem (evaluations)  \
0                                        6.0   
1                                        0.0   
2                                        8.0   
3                                        7.0   
4                                        NaN   
...                                      ...   
3091                                     9.0   
3092             

**3️. Tampilkan tipe data dari setiap kolom yang terdapat dalam dataset.**

In [ ]:
# Menampilkan tipe data dari setiap kolom pada dataset hasil gabungan
data.dtypes


Unnamed: 0                                          int64
Student Id                                          int64
Marital status                                     object
Application mode                                  float64
Application order                                 float64
Course                                            float64
Daytime/evening attendance\t                       object
Previous qualification                            float64
Previous qualification (grade)                    float64
Nacionality                                        object
Mother's qualification                            float64
Father's qualification                            float64
Mother's occupation                               float64
Father's occupation                               float64
Admission grade                                   float64
Displaced                                         float64
Educational special needs                         float64
Debtor                                            float64
Tuition fees up to date                           float64
Gender                                            float64
Scholarship holder                                float64
Age at enrollment                                 float64
International                                     float64
Curricular units 1st sem (credited)               float64
Curricular units 1st sem (enrolled)               float64
Curricular units 1st sem (evaluations)            float64
Curricular units 1st sem (approved)               float64
Curricular units 1st sem (grade)                  float64
Curricular units 1st sem (without evaluations)    float64
Curricular units 2nd sem (credited)               float64
Curricular units 2nd sem (enrolled)               float64
Curricular units 2nd sem (evaluations)            float64
Curricular units 2nd sem (approved)               float64
Curricular units 2nd sem (grade)                  float64
Curricular units 2nd sem (without evaluations)    float64
Unemployment rate                                 float64
Inflation rate                                    float64
GDP                                               float64
Target                                             object
dtype: object

**4. Tampilkan statistik deskriptif dari dataset dan berikan analisis singkat mengenai hasilnya.**


In [ ]:
# Menampilkan statistik deskriptif dari dataset
data.describe().T


count         mean  \
Unnamed: 0                                      3096.0  1547.500000   
Student Id                                      3096.0  5470.623062   
Application mode                                2924.0    18.785568   
Application order                               2988.0     1.734940   
Course                                          2944.0  8817.929348   
Previous qualification                          2973.0     4.685503   
Previous qualification (grade)                  3011.0   132.647293   
Mother's qualification                          2951.0    19.470010   
Father's qualification                          3000.0    22.408667   
Mother's occupation                             3034.0    11.495715   
Father's occupation                             2952.0    11.597900   
Admission grade                                 2955.0   126.902876   
Displaced                                       2974.0     0.546066   
Educational special needs                       2977.0     0.010749   
Debtor                                          2980.0     0.117450   
Tuition fees up to date                         2940.0     0.880612   
Gender                                          2894.0     0.350726   
Scholarship holder                              2930.0     0.246416   
Age at enrollment                               2892.0    23.310166   
International                                   2995.0     0.025042   
Curricular units 1st sem (credited)             2888.0     0.698407   
Curricular units 1st sem (enrolled)             2953.0     6.251270   
Curricular units 1st sem (evaluations)          3029.0     8.233080   
Curricular units 1st sem (approved)             2912.0     4.732143   
Curricular units 1st sem (grade)                3007.0    10.692416   
Curricular units 1st sem (without evaluations)  3023.0     0.133311   
Curricular units 2nd sem (credited)             3035.0     0.516639   
Curricular units 2nd sem (enrolled)             2997.0     6.208876   
Curricular units 2nd sem (evaluations)          2889.0     8.061613   
Curricular units 2nd sem (approved)             2945.0     4.422411   
Curricular units 2nd sem (grade)                3021.0    10.275115   
Curricular units 2nd sem (without evaluations)  2943.0     0.139314   
Unemployment rate                               2960.0    11.549561   
Inflation rate                                  2929.0     1.215534   
GDP                                             3032.0     0.028278   

                                                        std      min      25%  \
Unnamed: 0                                       893.882543     0.00   773.75   
Student Id                                      2582.483019  1000.00  3288.50   
Application mode                                  17.532674     1.00     1.00   
Application order                                  1.327527     1.00     1.00   
Course                                          2148.407432    33.00  9085.00   
Previous qualification                            10.439709     1.00     1.00   
Previous qualification (grade)                    13.322662    95.00   124.00   
Mother's qualification                            15.636952     1.00     2.00   
Father's qualification                            15.291008     1.00     3.00   
Mother's occupation                               28.173386     0.00     4.00   
Father's occupation                               26.799023     0.00     4.00   
Admission grade                                   14.460283    95.00   118.00   
Displaced                                          0.497957     0.00     0.00   
Educational special needs                          0.103136     0.00     0.00   
Debtor                                             0.322009     0.00     0.00   
Tuition fees up to date                            0.324299     0.00     1.00   
Gender                                             0.477280     0.00     0.00   
Scholarship holder                

Analisis Statistik Deskriptif Dataset Mahasiswa

Age at enrollment (Usia Saat Mendaftar)

Rata-rata usia mahasiswa saat mendaftar adalah sekitar 23 tahun dengan nilai minimum 17 tahun dan maksimum 70 tahun.

Nilai standar deviasi yang cukup tinggi menunjukkan adanya variasi usia yang lebar antar mahasiswa.

Sebagian besar mahasiswa (sekitar 75%) berusia di bawah 25 tahun, menandakan bahwa mayoritas merupakan mahasiswa muda yang baru lulus dari pendidikan sebelumnya.

Admission grade (Nilai Seleksi Masuk)

Nilai rata-rata seleksi masuk sekitar 125, dengan rentang dari 95 hingga 190.

Nilai median 125 menunjukkan bahwa distribusi nilai cukup seimbang.

Nilai maksimum 190 bisa jadi merupakan outlier, menunjukkan mahasiswa dengan performa akademik yang sangat tinggi pada saat seleksi.

Curricular units 1st sem (approved)

Rata-rata mahasiswa menyelesaikan sekitar 4–5 mata kuliah pada semester pertama dengan standar deviasi sekitar 3.0, menandakan adanya variasi performa belajar antar mahasiswa.

Ada mahasiswa yang belum lulus satupun mata kuliah (nilai minimum = 0), sementara ada juga yang lulus hingga 21 mata kuliah, menunjukkan perbedaan kemampuan dan beban studi yang signifikan.

Curricular units 2nd sem (grade)

Nilai rata-rata mahasiswa pada semester kedua adalah sekitar 10.8, dengan nilai maksimum 18.5.

Sebagian besar mahasiswa memiliki nilai sekitar 12–13, menunjukkan performa yang cukup baik dan stabil antar semester.

Unemployment rate dan GDP

Rata-rata tingkat pengangguran pada data ini adalah 11.5%, dengan variasi yang cukup tinggi antar periode.

Nilai GDP rata-rata sekitar 0.028, menunjukkan kondisi ekonomi yang fluktuatif pada masa pengumpulan data.

Adanya hubungan potensial antara faktor ekonomi (GDP dan unemployment rate) dengan performa mahasiswa bisa menjadi bahan analisis lanjutan.

Kesimpulan Umum

Dataset menunjukkan keragaman karakteristik mahasiswa yang cukup besar, baik dari segi usia, prestasi akademik, maupun faktor ekonomi.

Beberapa kolom memiliki outlier atau nilai ekstrim (seperti usia 70 tahun dan nilai 190) yang perlu diperhatikan sebelum melakukan modeling.

Distribusi data secara umum tidak simetris (skewed), sehingga transformasi data atau normalisasi mungkin diperlukan pada tahap pra-pemrosesan.

Secara keseluruhan, data sudah cukup informatif untuk mendukung analisis prediktif terkait performa mahasiswa.

**5️. Slicing Data**

In [ ]:
df.head(20)

Unnamed: 0  Student Id Marital status  Application mode  \
0            0        6634         single               1.0   
1            1        6459         single               1.0   
2            2        2238         single              42.0   
3            3        2479         single               1.0   
4            4        4260         single              44.0   
5            5        3544        married              39.0   
6            6        4956         single              39.0   
7            7        6758         single               1.0   
8            8        2484         single               1.0   
9            9        9845         single              39.0   
10          10        3690         single               1.0   
11          11        2585         single               1.0   
12          12        5241         single               1.0   
13          13        2972         single               1.0   
14          14        2003        married              39.0   
15          15        3007         single               7.0   
16          16        9857         single              17.0   
17          17        3327         single               NaN   
18          18        5235         single              17.0   
19          19        3969         single               1.0   

    Application order  Course Daytime/evening attendance\t  \
0                 3.0  9773.0                          day   
1                 1.0     NaN                          day   
2                 1.0  9500.0                          day   
3                 1.0     NaN                          day   
4                 1.0     NaN                          day   
5                 1.0  9853.0                          day   
6                 1.0  9085.0                          day   
7                 4.0  9670.0                          day   
8                 1.0  9773.0                          day   
9                 1.0  9670.0                          day   
10                1.0  9670.0                          day   
11                1.0  9070.0                          day   
12                1.0   171.0                          day   
13                3.0  9500.0                          day   
14                1.0  8014.0                          NaN   
15                1.0  9119.0                          day   
16                1.0  9119.0                          day   
17                1.0  9003.0                          day   
18                1.0   171.0                          day   
19                1.0  9085.0                          day   

    Previous qualification  Previous qualification (grade) Nacionality  ...  \
0                      1.0                             NaN  Portuguese  ...   
1                      1.0                           136.0  Portuguese  ...   
2                      1.0                           120.0  Portuguese  ...   
3                      1.0                           141.0  Portuguese  ...   
4                     39.0                           120.0     Spanish  ...   
5                      1.0                             NaN  Portuguese  ...   
6                      1.0                           133.1         NaN  ...   
7                      1.0                           130.0  Portuguese  ...   
8                      1.0                           135.0  Portuguese  ...   
9                      1.0                           120.0  Portuguese  ...   
10                     1.0                           133.1  Portuguese  ...   
11                     1.0                           117.0  Portuguese  ...   
12                     1.0                           162.0  Portuguese  ...   
13                     1.0                           135.0  Portuguese  ...   
14                     NaN                           120.0  Portuguese  ...   
15                     1.0                           140.8  Portuguese  ...   
16                     1.0        

In [ ]:
df.tail(20)

Unnamed: 0  Student Id Marital status  Application mode  \
3076        3076        6757         single              17.0   
3077        3077        5750         single               1.0   
3078        3078        2599         single               1.0   
3079        3079        3249         single              17.0   
3080        3080        3366       divorced              39.0   
3081        3081        3462         single              39.0   
3082        3082        9194         single              17.0   
3083        3083        5574         single               1.0   
3084        3084        8311         single               1.0   
3085        3085        2304         single              17.0   
3086        3086        6615         single              43.0   
3087        3087        3759         single               1.0   
3088        3088        8085         single              51.0   
3089        3089        1573         single              17.0   
3090        3090        9737         single              51.0   
3091        3091        6176         single              44.0   
3092        3092        8064         single               1.0   
3093        3093        2103         single               1.0   
3094        3094        8629         single              39.0   
3095        3095        6614         single              17.0   

      Application order  Course Daytime/evening attendance\t  \
3076                NaN  9254.0                          day   
3077                2.0  9500.0                          day   
3078                1.0  9085.0                          day   
3079                1.0  9238.0                          day   
3080                1.0  9003.0                          day   
3081                1.0   171.0                          day   
3082                5.0  9147.0                          day   
3083                1.0  9500.0                          day   
3084                2.0  9773.0                          day   
3085                1.0  9238.0                          day   
3086                1.0  9556.0                          day   
3087                1.0   171.0                          day   
3088                1.0  9670.0                          day   
3089                2.0   171.0                          day   
3090                1.0  9070.0                          day   
3091                1.0  9085.0                          day   
3092                5.0  9238.0                          day   
3093                2.0  9500.0                          day   
3094                1.0  9500.0                          day   
3095                1.0  9670.0                          day   

      Previous qualification  Previous qualification (grade) Nacionality  ...  \
3076                     1.0                           122.0  Portuguese  ...   
3077                     1.0                           140.0  Portuguese  ...   
3078                     1.0                           131.0  Portuguese  ...   
3079                     1.0                           120.0  Portuguese  ...   
3080                    12.0                           133.1  Portuguese  ...   
3081                     1.0                           147.0  Portuguese  ...   
3082                     1.0                           131.0  Portuguese  ...   
3083                     1.0                           140.0  Portuguese  ...   
3084                     1.0                           130.0  Portuguese  ...   
3085                     NaN                           139.0  Portuguese  ...   
3086                     1.0                           149.0  Portuguese  ...   
3087                     3.0                           116.0  Portuguese  ...   
3088                     1.0                           133.1  Portuguese  ...   
3089                     1.0                           142.0  Portuguese  ...   
3090                     1.0                           143.0  Portuguese  ...   

In [ ]:
df.iloc[80:101]

Unnamed: 0  Student Id Marital status  Application mode  \
80           80        1238         single              17.0   
81           81        4121         single              39.0   
82           82        2890         single               1.0   
83           83        4358         single               1.0   
84           84        4896         single               NaN   
85           85        4735         single              18.0   
86           86        1047         single               1.0   
87           87        7430         single              39.0   
88           88        8527         single               1.0   
89           89        5420         single              43.0   
90           90        5625         single              44.0   
91           91        8990            NaN               1.0   
92           92        9143        married              44.0   
93           93        1944        married               7.0   
94           94        3666         single               1.0   
95           95        6960         single              17.0   
96           96        7620         single               1.0   
97           97        6083         single              42.0   
98           98        5525         single              17.0   
99           99        6029         single              17.0   
100         100        9749         single              39.0   

     Application order  Course Daytime/evening attendance\t  \
80                 1.0  9670.0                          day   
81                 1.0  9003.0                          day   
82                 1.0   171.0                          day   
83                 1.0  9147.0                          day   
84                 1.0  9991.0                      evening   
85                 1.0  9500.0                          day   
86                 1.0  9773.0                          day   
87                 2.0  9556.0                          day   
88                 2.0  9500.0                          day   
89                 1.0     NaN                          day   
90                 1.0  9003.0                          day   
91                 1.0  9500.0                          day   
92                 1.0  9085.0                          day   
93                 1.0  9070.0                          day   
94                 3.0  9070.0                          day   
95                 3.0  9119.0                          day   
96                 1.0  9500.0                          day   
97                 1.0  9500.0                          day   
98                 5.0  9238.0                          day   
99                 1.0  9773.0                          day   
100                1.0  9085.0                          day   

     Previous qualification  Previous qualification (grade) Nacionality  ...  \
80                      1.0                           155.0  Portuguese  ...   
81                      1.0                             NaN  Portuguese  ...   
82                      1.0                           159.0         NaN  ...   
83                      1.0                           150.0   Ukrainian  ...   
84                      1.0                           140.0  Portuguese  ...   
85                      1.0                           144.0  Portuguese  ...   
86                      1.0                           117.0  Portuguese  ...   
87                     19.0                           133.1  Portuguese  ...   
88                      1.0                           145.0  Portuguese  ...   
89                      1.0                           121.0  Portuguese  ...   
90                     39.0                           140.0  Portuguese  ...   
91                      1.0                           141.0  Portuguese  ...   
92                     39.0                           160.0  Portuguese  ...   
93                     43.0                           140.0  Portuguese  ...   
94       

**6️. Tangani nilai yang hilang (NaN) menggunakan salah satu metode berikut:**

In [ ]:

# Mengecek jumlah nilai hilang di setiap kolom
print("=== Jumlah Nilai Hilang per Kolom ===")
print(df.isnull().sum())

# Menghitung total keseluruhan nilai hilang
print("\nTotal nilai hilang di seluruh dataset:", df.isnull().sum().sum())

# Menghapus baris yang seluruh kolomnya kosong (jika ada)
df = df.dropna(how='all')

# Mengisi nilai hilang dengan rata-rata pada kolom numerik
df = df.fillna(df.mean(numeric_only=True))

# Mengecek ulang apakah masih ada nilai hilang
print("\n=== Setelah Penanganan ===")
print(df.isnull().sum())

=== Jumlah Nilai Hilang per Kolom ===
Unnamed: 0                                          0
Student Id                                          0
Marital status                                     84
Application mode                                  172
Application order                                 108
Course                                            152
Daytime/evening attendance\t                       80
Previous qualification                            123
Previous qualification (grade)                     85
Nacionality                                       185
Mother's qualification                            145
Father's qualification                             96
Mother's occupation                                62
Father's occupation                               144
Admission grade                                   141
Displaced                                         122
Educational special needs                         119
Debtor                                      

**7️. Opsional : Lakukan encoding terhadap kolom bertipe object menjadi numerik menggunakan:**

In [ ]:

# Menampilkan kolom dengan tipe data object (kategori)
print("=== Kolom Kategorikal ===")
kategori_cols = df.select_dtypes(include=['object']).columns
print(kategori_cols)

# Melakukan encoding one-hot untuk kolom kategorikal
df_encoded = pd.get_dummies(df, columns=kategori_cols, drop_first=True)

# Menampilkan hasil encoding
print("\nData setelah encoding:")
display(df_encoded.head())

# Mengecek ukuran data setelah encoding
print("\nUkuran data setelah encoding:", df_encoded.shape)

=== Kolom Kategorikal ===
Index(['Marital status', 'Daytime/evening attendance\t', 'Nacionality',
       'Target'],
      dtype='object')

Data setelah encoding:


Unnamed: 0  Student Id  Application mode  Application order       Course  \
0           0        6634               1.0                3.0  9773.000000   
1           1        6459               1.0                1.0  8817.929348   
2           2        2238              42.0                1.0  9500.000000   
3           3        2479               1.0                1.0  8817.929348   
4           4        4260              44.0                1.0  8817.929348   

   Previous qualification  Previous qualification (grade)  \
0                     1.0                      132.647293   
1                     1.0                      136.000000   
2                     1.0                      120.000000   
3                     1.0                      141.000000   
4                    39.0                      120.000000   

   Mother's qualification  Father's qualification  Mother's occupation  ...  \
0                     1.0                     1.0                  4.0  ...   
1                    19.0                    19.0                  4.0  ...   
2                     1.0                     2.0                  9.0  ...   
3                     1.0                    19.0                  4.0  ...   
4                     4.0                     4.0                  3.0  ...   

   Nacionality_Moldova (Republic of)  Nacionality_Mozambican  \
0                              False                   False   
1                              False                   False   
2                              False                   False   
3                              False                   False   
4                              False                   False   

   Nacionality_Portuguese  Nacionality_Romanian  Nacionality_Russian  \
0                    True                 False                False   
1                    True                 False                False   
2                    True                 False                False   
3                    True                 False                False   
4                   False                 False                False   

   Nacionality_Santomean  Nacionality_Spanish  Nacionality_Ukrainian  \
0                  False                False                  False   
1                  False                False                  False   
2                  False                False                  False   
3                  False                False                  False   
4                  False                 True                  False   

   Target_Enrolled  Target_Graduate  
0            False             True  
1            False             True  
2            False             True  
3            False             True  
4            False            False  

[5 rows x 61 columns]


Ukuran data setelah encoding: (3096, 61)


**8️. Tampilkan kembali dataset yang telah melalui seluruh proses preprocessing (cleaning, imputasi, dan encoding).**

In [ ]:
from sklearn.preprocessing import StandardScaler

# Menyalin dataset hasil encoding agar aman
df_scaled = df_encoded.copy()

# Memilih kolom numerik
num_cols = df_scaled.select_dtypes(include=['int64', 'float64']).columns

# Membuat objek scaler
scaler = StandardScaler()

# Melakukan standarisasi pada kolom numerik
df_scaled[num_cols] = scaler.fit_transform(df_scaled[num_cols])

# Menampilkan hasil normalisasi
print("=== Data Setelah Normalisasi ===")
display(df_scaled.head())

# Mengecek ukuran dataset akhir
print("\nUkuran data setelah normalisasi:", df_scaled.shape)

=== Data Setelah Normalisasi ===


Unnamed: 0  Student Id  Application mode  Application order    Course  \
0   -1.731491    0.450561         -1.044012           0.970177  0.455957   
1   -1.730373    0.382785         -1.044012          -0.563626  0.000000   
2   -1.729254   -1.251952          1.362687          -0.563626  0.325625   
3   -1.728135   -1.158616         -1.044012          -0.563626  0.000000   
4   -1.727016   -0.468858          1.480087          -0.563626  0.000000   

   Previous qualification  Previous qualification (grade)  \
0               -0.360317                        0.000000   
1               -0.360317                        0.255224   
2               -0.360317                       -0.962773   
3               -0.360317                        0.635848   
4                3.354790                       -0.962773   

   Mother's qualification  Father's qualification  Mother's occupation  ...  \
0               -1.210053               -1.422544            -0.268806  ...   
1               -0.030792               -0.226496            -0.268806  ...   
2               -1.210053               -1.356097            -0.089499  ...   
3               -1.210053               -0.226496            -0.268806  ...   
4               -1.013510               -1.223203            -0.304667  ...   

   Nacionality_Moldova (Republic of)  Nacionality_Mozambican  \
0                              False                   False   
1                              False                   False   
2                              False                   False   
3                              False                   False   
4                              False                   False   

   Nacionality_Portuguese  Nacionality_Romanian  Nacionality_Russian  \
0                    True                 False                False   
1                    True                 False                False   
2                    True                 False                False   
3                    True                 False                False   
4                   False                 False                False   

   Nacionality_Santomean  Nacionality_Spanish  Nacionality_Ukrainian  \
0                  False                False                  False   
1                  False                False                  False   
2                  False                False                  False   
3                  False                False                  False   
4                  False                 True                  False   

   Target_Enrolled  Target_Graduate  
0            False             True  
1            False             True  
2            False             True  
3            False             True  
4            False            False  

[5 rows x 61 columns]


Ukuran data setelah normalisasi: (3096, 61)
